File: src/go/3_StructuralPatterns/2_Bridge/bridge.go
package structural_bridge

import "fmt"

type Renderer interface {
  RenderCircle(radius float32)
}

type VectorRenderer struct {

}

func (v *VectorRenderer) RenderCircle(radius float32) {
  fmt.Println("Drawing a circle of radius", radius)
}

type RasterRenderer struct {
  Dpi int
}

func (r *RasterRenderer) RenderCircle(radius float32) {
  fmt.Println("Drawing pixels for circle of radius", radius)
}

type Circle struct {
  renderer Renderer
  radius float32
}

func (c *Circle) Draw() {
  c.renderer.RenderCircle(c.radius)
}

func NewCircle(renderer Renderer, radius float32) *Circle {
  return &Circle{renderer: renderer, radius: radius}
}

func (c *Circle) Resize(factor float32) {
  c.radius *= factor
}

func main() {
  raster := RasterRenderer{}
  vector := VectorRenderer{}
  circle := NewCircle(&vector, 5)
  circle.Draw()
}

File: src/java/3_StructuralPatterns/2_Bridge/Bridge.java
package com.activemesa.structural.bridge;

import com.google.inject.AbstractModule;
import com.google.inject.Guice;
import com.google.inject.Inject;
import com.google.inject.Injector;

interface Renderer
{
  void renderCircle(float radius);
}

class VectorRenderer implements Renderer
{
  @Override
  public void renderCircle(float radius)
  {
    System.out.println("Drawing a circle of radius " + radius);
  }
}

class RasterRenderer implements Renderer
{
  @Override
  public void renderCircle(float radius)
  {
    System.out.println("Drawing pixels for a circle of radius " + radius);
  }
}

abstract class Shape
{
  protected Renderer renderer;

  public Shape(Renderer renderer)
  {
    this.renderer = renderer;
  }

  public abstract void draw();
  public abstract void resize(float factor);
}

class Circle extends Shape
{
  public float radius;

  @Inject
  public Circle(Renderer renderer)
  {
    super(renderer);
  }

  public Circle(Renderer renderer, float radius)
  {
    super(renderer);
    this.radius = radius;
  }

  @Override
  public void draw()
  {
    renderer.renderCircle(radius);
  }

  @Override
  public void resize(float factor)
  {
    radius *= factor;
  }
}

class ShapeModule extends AbstractModule
{
  @Override
  protected void configure()
  {
    bind(Renderer.class).to(VectorRenderer.class);
  }
}

class BridgeDemo
{
  public static void main(String[] args)
  {
//    RasterRenderer rasterRenderer = new RasterRenderer();
//    VectorRenderer vectorRenderer = new VectorRenderer();
//    Circle circle = new Circle(vectorRenderer, 5);
//    circle.draw();
//    circle.resize(2);
//    circle.draw();

    // todo: Google Guice
    Injector injector = Guice.createInjector(new ShapeModule());
    Circle instance = injector.getInstance(Circle.class);
    instance.radius = 3;
    instance.draw();
    instance.resize(2);
    instance.draw();
  }
}
